Some benchmark improvements.

shawjef3 opened the pull request:
Move logger acquisition out of the benchmark methods and into a Setup method.

Enable async logging for log4j2. I thought just including the Disruptor jar was enough to do that, but it's not. 14x speedup for regular logger and 64x speedup for trace.

Before and after, which I ran with `jmh:run -i 5 -wi 10 -f1 -t1 -rf JSON -rff benchmarks.json`:

> [info] # Run complete. Total time: 00:02:32
[info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    5   7385.325 ±  1648.339  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    5  61592.498 ± 28139.187  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    5  38124.842 ±  7714.751  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    5  44042.671 ±  3838.535  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    5  45312.231 ± 33704.914  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    5  52173.400 ± 20511.603  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    5   1000.674 ±    29.562  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    5     73.252 ±     2.777  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    5  10483.946 ±  1050.459  ns/op
>
> [info] # Run complete. Total time: 00:02:32
[info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    5    524.290 ±    18.433  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    5    959.314 ±   110.947  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    5  39506.749 ±  5119.532  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    5  41633.737 ±  2861.170  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    5  39028.793 ± 14129.436  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    5  36399.867 ±  4559.278  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    5    955.418 ±    59.406  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    5     70.282 ±     1.380  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    5  10586.169 ±  1085.722  ns/op

darkfrog26 commented:
I get these results when I run your code:

```
[info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    3    350.942 ±   188.742  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    3    610.531 ±   296.220  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    3  20995.516 ±  3884.130  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    3  22410.248 ±  6468.598  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    3   6953.548 ±  1106.238  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    3  20559.968 ±  4075.373  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    3   1241.676 ± 19049.277  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    3     53.240 ±     3.188  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    3   3741.872 ±   126.071  ns/op
```

When I run against master I get:

```
[info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    3   1700.685 ±   412.293  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    3  29362.470 ±  4246.313  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    3  20142.659 ±   964.661  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    3  22085.284 ±   875.400  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    3   6975.190 ± 18676.463  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    3  20283.890 ±  1379.827  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    3    642.788 ±   797.574  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    3     57.849 ±     2.543  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    3   3702.153 ±   272.608  ns/op
```

While it's clear you've optimized log4j, Scribe is going MUCH slower with your configuration changes.

darkfrog26 commented:
If I just incorporate your `log4j2.xml` changes:

```
[info] Benchmark                               Mode  Cnt      Score      Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    3    419.053 ±    3.255  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    3  25517.299 ±  188.848  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    3  20029.900 ± 1680.708  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    3  22024.577 ± 2399.345  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    3   6747.899 ± 1485.220  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    3  20409.795 ±  700.101  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    3    586.211 ±  175.799  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    3     57.262 ±    3.536  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    3   3709.096 ±  257.617  ns/op
```

darkfrog26 commented:
I also believe you've screwed up the configuration for log4j in your changes:
![screenshot from 2018-06-28 13-42-08](https://user-images.githubusercontent.com/877534/42054123-24f0755a-7ad9-11e8-85e5-a4a56b888f13.png)

darkfrog26 commented:
It's a lot easier to be faster when you're not actually doing anything. ;)

darkfrog26 commented:
This is what it looks like with the original `log4j2.xml` file:

![screenshot from 2018-06-28 13-46-18](https://user-images.githubusercontent.com/877534/42054355-b99f54be-7ad9-11e8-97d5-e2b4fb7629de.png)

Although, I just noticed that `log4s.log` doesn't appear to be writing. Crazy how slow it is for not doing anything...

shawjef3 commented:
I fixed the log4j2 benchmarks, but not the log4s one.

> [info] # Run complete. Total time: 00:02:34
[info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    5    536.393 ±     7.125  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    5    985.039 ±   358.063  ns/op
[info] LoggingSpeedBenchmark.withLog4s         avgt    5  42153.856 ± 28214.769  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    5  42327.786 ±  3374.168  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    5  28388.631 ±  6614.489  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    5  40789.176 ± 17002.872  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    5    951.378 ±   114.443  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    5     69.512 ±     1.483  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    5  10117.219 ±   384.815  ns/op

> $ ls -lh benchmarks/logs
total 46148776
-rw-r--r--  1 jshaw  staff   1.2G Jun 28 16:39 log4j-trace.log
-rw-r--r--  1 jshaw  staff   8.0G Jun 28 16:39 log4j.log
-rw-r--r--  1 jshaw  staff     0B Jun 28 11:47 log4s.log
-rw-r--r--  1 jshaw  staff    62M Jun 28 16:39 logback.log
-rw-r--r--  1 jshaw  staff   533M Jun 28 16:40 scala-logging.log
-rw-r--r--  1 jshaw  staff    11M Jun 28 16:41 scribe-async.log
-rw-r--r--  1 jshaw  staff    12G Jun 28 16:40 scribe.log
-rw-r--r--  1 jshaw  staff   210M Jun 28 16:41 tiny.log

shawjef3 commented:
I also removed scala-logging, since that's also just a facade over slf4j.

I added a benchmark for asynchronous logback appenders, which I didn't know existed. Logback becomes very fast at returning control back to the thread, but its performance in terms of messages/second is poor.

> [info] Benchmark                               Mode  Cnt      Score       Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    5    595.182 ±   212.298  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    5   1091.724 ±    58.610  ns/op
[info] LoggingSpeedBenchmark.withLogback       avgt    5  35274.435 ±  1491.399  ns/op
[info] LoggingSpeedBenchmark.withLogbackAsync  avgt    5     74.962 ±     0.964  ns/op
[info] LoggingSpeedBenchmark.withPrintLine     avgt    5  33606.637 ± 11325.569  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    5    980.842 ±   283.511  ns/op
[info] LoggingSpeedBenchmark.withScribeAsync   avgt    5     69.882 ±     1.968  ns/op
[info] LoggingSpeedBenchmark.withTinyLog       avgt    5   9764.815 ±   282.540  ns/op
[info] Benchmark result is saved to benchmarks.json
[success] Total time: 138 s, completed Jun 28, 2018 9:54:04 PM

> $ ls -lh benchmarks/logs
total 15706568
-rw-r--r--  1 jshaw  staff   1.4G Jun 28 21:52 log4j-trace.log
-rw-r--r--  1 jshaw  staff   3.9G Jun 28 21:52 log4j.log
-rw-r--r--  1 jshaw  staff    79M Jun 28 21:52 logback-async.log
-rw-r--r--  1 jshaw  staff    64M Jun 28 21:52 logback.log
-rw-r--r--  1 jshaw  staff   1.8M Jun 28 21:53 scribe-async.log
-rw-r--r--  1 jshaw  staff   1.8G Jun 28 21:53 scribe.log
-rw-r--r--  1 jshaw  staff   222M Jun 28 21:54 tiny.log

darkfrog26 commented:
You have to commit your changes to the branch if I'm going to be able to see it.

darkfrog26 commented:
@shawjef3, if you still want this merged you need to update it with your latest changes

darkfrog26 commented:
@shawjef3, I'm still waiting for an update on this.  If you're not interested in following up let me know and I'll close the PR.
